downstairs/src/extent.rs

```diff
@@ -0,0 +1,829 @@
+         * there are not too many files in any level of that hierarchy.
+         */
+        let path = extent_path(dir, number);
+
+        // If there was an unfinished copy, then clean it up here (??)
+        remove_copy_cleanup_dir(dir, number)?;
+
+        // If the replace directory exists for this extent, then it means
+        // a repair was interrupted before it could finish.  We will continue
+        // the repair before we open the extent.
+        let replace_dir = replace_dir(dir, number);
+        if !read_only && Path::new(&replace_dir).exists() {
+            info!(
+                log,
+                "Extent {} found replacement dir, finishing replacement",
+                number
+            );
+            move_replacement_extent(dir, number as usize, log)?;
+        }
+
+        // Pick the format for the downstairs files
+        //
+        // Right now, this only supports SQLite-flavored Downstairs
+        let inner = {
+            let mut sqlite_path = path.clone();
+            sqlite_path.set_extension("db");
+            if sqlite_path.exists() {
+                let inner = extent_inner_sqlite::SqliteInner::open(
+                    &path, def, number, read_only, log,
+                )?;
+                Box::new(inner)
+            } else {
+                panic!("no SQLite file present at {sqlite_path:?}");
+            }
+        };
+
+        // XXX: schema updates?
+
+        let extent = Extent {
+            number,
+            read_only,
+            iov_max: Extent::get_iov_max()?,
+            inner: Mutex::new(inner),
+        };
+
+        Ok(extent)
+    }
+
+    pub async fn dirty(&self) -> bool {
+        self.inner.lock().await.dirty().unwrap()
+    }
+
+    /**
+     * Close an extent and the metadata db files for it.
+     */
+    pub async fn close(self) -> Result<(u64, u64, bool), CrucibleError> {
+        let inner = self.inner.lock().await;
+
+        let gen = inner.gen_number().unwrap();
+        let flush = inner.flush_number().unwrap();
+        let dirty = inner.dirty().unwrap();
+
+        Ok((gen, flush, dirty))
+    }
+
+    /**
+     * Create an extent at the location requested.
+     * Start off with the default meta data.
+     * Note that this function is not safe to run concurrently.
+     */
+    pub fn create(
+        dir: &Path,
+        def: &RegionDefinition,
+        number: u32,
+    ) -> Result<Extent> {
+        /*
+         * Store extent data in files within a directory hierarchy so that
+         * there are not too many files in any level of that hierarchy.
+         */
+        {
+            let path = extent_path(dir, number);
+
+            /*
+             * Verify there are not existing extent files.
+             */
+            if Path::new(&path).exists() {
+                bail!("Extent file already exists {:?}", path);
+            }
+        }
+        remove_copy_cleanup_dir(dir, number)?;
```

Comment:
Maybe this is a comment for the existing code and not this PR, but is this needed here?

Reply:
This is a [preexisting condition](https://github.com/oxidecomputer/crucible/blame/main/downstairs/src/region.rs#L771).  I agree that it would be a little funky if someone created a extent into a directory that had a left-over copy file, but that _could_ be possible?  Anyways, I think it's harmless.